Phase 3: Documentation (#142)
Ekumen-OS/lambkin: src/lambkin/README.md

Reviewer: My recommendation would be a short snippet in the SDK README, with a pointer to Beluga for the full picture. Something that shows how you import the SDK, define the three stages, wire them together, and kick off a run.

The reason not to rely solely on Beluga for this is that Beluga carries a lot of weight: Docker, ROS2, rosbags, a specific algorithm. A reader who just wants to understand the SDK's shape has to wade through all of that to find the structural pattern. A skeleton example in the SDK README removes that barrier.

Author: Solved in 9698aff

# LAMBKIN SDK

The LAMBKIN Python SDK is the core library for building SLAM evaluation pipelines. It provides the orchestration primitives you compose into your benchmark: process lifecycle management, parameter sweep execution, pipeline stages, and structured I/O.

## Architecture

A benchmark is structured around three stages that LAMBKIN sequences and keeps organized:

1. **Ingestion** — locate or download datasets and prepare inputs.
2. **Execution** — launch and supervise the processes that run your algorithm.
3. **Egression** — collect outputs and compute metrics.

Each stage is a Python callable that receives a context object carrying configuration, paths, and state. You implement the logic; LAMBKIN handles the rest.

## Core Concepts

LAMBKIN exposes a small set of composable primitives. Together they cover the full lifecycle of a benchmark — from declaring inputs and sweeping parameters to launching processes and collecting results.

> **`named_product`**

Reviewer: To follow the same pattern as the rest of core concepts:
```suggestion
**`Named Product`**
```

Author: Solved in d727ae8

Takes named parameter lists and returns every possible combination as a list of dictionaries, one per benchmark run configuration. Pass the result to `Benchmark` via `variants=` to sweep all combinations automatically.

**`Benchmark`**
Drives the benchmark execution loop. It parses options registered via `Option` once before the loop, then creates a `Context` for every `(variant, iteration)` pair and calls the decorated function with it. A base context is used during setup to resolve inputs before the loop begins. Can be used as a decorator via `@benchmark`.

**`Context`**

> Carries all namespaced information for one benchmark variant. Builds `ctx.variant`, `ctx.inputs`, `ctx.options`, and `ctx.output` from the given parameters, and automatically creates the required output folders on disk before the benchmark function runs.

Reviewer: "Carries all namespaced information for one benchmark variant". Variant and **iteration** also no? 

Still seeing implementation details here via `ctx.<name>`. These details regarding class members and even attributes is more API documentation, not so relevant for a README explaining an overview of the core concepts.

Author: Solved in 75c4873

**`Source`**
Describes the benchmark script being executed. Exposed on the context as `ctx.source`, it gives benchmark stages access to the script's location and metadata without hardcoding paths.

**`Input`**

> Registers a data resolution hook on a benchmark. Hooks are resolved before each iteration and their return values injected into `ctx.inputs` under the hook's function name, keeping data resolution decoupled from benchmark logic. Can be used as a decorator via `@input`.

Reviewer: ```suggestion
Registers a data resolution hook on a benchmark. Hooks are resolved before each iteration and their return values injected into the `Context` under the hook's function name, keeping data resolution decoupled from benchmark logic. Can be used as a decorator via `@input`.
```

Reviewer: Missing the explanation that they require a specific function signature and that names must be unique.

Author: Solved in e0eb2d9

**`Option`**
Abstracts shell command dispatch. Exposes the host environment's executables as Python attributes — accessing `shell.my_tool` returns a callable that runs `my_tool` with the given arguments, letting benchmark scripts invoke external processes without hardcoding paths or constructing subprocess calls manually. Accessible via `ctx.shell`.

**`Shell`**

> Exposes the host environment's executables as Python attributes. Accessing `ctx.shell.my_tool` returns a callable that, when invoked, runs `my_tool` with the given arguments. This lets benchmark scripts call external processes as if they were native Python functions, without hardcoding paths or constructing subprocess calls manually.

Reviewer: I think something got messed up here. Option is not right. Please review.

Author: Solved in eb8c731

## Requirements

- Python 3.10+
- [`uv`](https://github.com/astral-sh/uv)

## Installation

```bash
git clone git@github.com:Ekumen-OS/lambkin.git
uv sync
```

## Usage

> A benchmark is a decorated Python function. The @lambkin.benchmark decorator handles iteration, parameter expansion, and context setup. Inputs and outputs are registered as hooks on the benchmark function.

Reviewer: ```suggestion
A benchmark is a decorated Python function. The `@lambkin.benchmark` decorator handles iteration, parameter expansion, and context setup. Inputs and outputs are registered as hooks on the benchmark function.
```

Reviewer: Maybe this info "The @lambkin.benchmark decorator handles iteration, parameter expansion, and context setup. Inputs and outputs are registered as hooks on the benchmark function." can be used to complement the core concepts section (in the appropriate subsections) and for usage just add a simple comment on the example.

Author: Solved in ef16705

```python
import lambkin

@lambkin.benchmark(
    variants=lambkin.common.named_product(
        param_a=["x", "y"],
        param_b=[1, 10, 100],
    ),
    num_iterations=5,
)
@lambkin.option("--clock-rate", default=100.0)
def my_benchmark(ctx):
    with lambkin.process.background(ctx.shell.my_recorder, "-O", "output.mcap"):
        with lambkin.process.background(
            ctx.shell.my_algorithm,
            f"param_a:={ctx.variant.param_a}",
            f"param_b:={ctx.variant.param_b}",
            f"input:={ctx.inputs.dataset}",
        ):
            ctx.shell.my_player(ctx.inputs.dataset, "-r", ctx.options.clock_rate)

@my_benchmark.input
def dataset(ctx):
    return ctx.source.path.parent / "datasets" / "my_dataset.mcap"

if __name__ == "__main__":
    my_benchmark()

```

> > **Note:** Real execution is not yet supported. `--dry-run` is the only supported mode at this time. The shell proxy prints commands rather than running them.

Reviewer: Add a new line in between the code and the note.

Author: Solved in d727ae8

# LAMBKIN

This is the LAMBKIN Python SDK — the core library for building SLAM evaluation pipelines. It provides the orchestration primitives you compose into your benchmark: process lifecycle management, parameter sweep execution, pipeline stages, and structured I/O.

> ---

Reviewer: I don't think the horizontal spacers are needed. The double quote already add a line underneath so IMO is cleanear without

Author: Solved in fdd54ae

## Architecture

### Output folder structure

LAMBKIN automatically creates the following directory structure for all generated artifacts:

```
results/
└── <variant_n>/
    └── iter_<n>/
        ├── output.mcap
        ├── out.zip
        └── ...
```

---

## Requirements

- Python 3.10+
- [`uv`](https://github.com/astral-sh/uv)

---

## Installation

Clone the repository and install the SDK dependencies:

```bash
git clone git@github.com:Ekumen-OS/lambkin.git
uv sync
```

---

## Getting started

You write a Python script that imports the SDK and composes its primitives into your benchmark pipeline. LAMBKIN handles the orchestration; you define the logic.

A benchmark is structured around three stages:

1. **Ingestion** — locate or download datasets and prepare inputs.
2. **Execution** — launch and supervise the processes that run your algorithm.

> 3. **Egression** — collect outputs and compute metrics.

Reviewer: Might be worth explaining this as part of how Lambkin works or is structured rather in a section that described how to get started. Could be misleading.

Author: Solved in 9698aff

# LAMBKIN SDK

The LAMBKIN Python SDK is the core library for building SLAM evaluation pipelines. It provides the orchestration primitives you compose into your benchmark: process lifecycle management, parameter sweep execution, pipeline stages, and structured I/O.

## Structure

> A benchmark is structured around three stages that LAMBKIN sequences, parallelizes, and keeps organized:

Reviewer: "parallelizes"? Same comment as in the other README

Author: Solved in ae67ef5

1. **Ingestion** — locate or download datasets and prepare inputs.
2. **Execution** — launch and supervise the processes that run your algorithm.
3. **Egression** — collect outputs and compute metrics.

Each stage is a Python callable that receives a context object carrying configuration, paths, and state. You implement the logic; LAMBKIN handles the rest.

LAMBKIN writes all artifacts under a consistent directory tree:
```
results/
└── <variant_n>/
    └── iter_<n>/
        ├── output.mcap
        ├── out.zip
        └── ...
```

## Usage

A benchmark is a decorated Python function. The `@lambkin.benchmark` decorator handles iteration, parameter expansion, and context setup.

> Inputs and outputs are registered as hooks on the benchmark function.

Reviewer: A comment on the shell would help understand better the SDK below and where `ctx.shell.my_recorde`r and `ctx.shell.my_algorithm` come from. No need to be super exhaustive but a quick and compact explanation to give some context would be useful for the reader to understand.

Author: Solved in ae67ef5

```python
import lambkin

@lambkin.benchmark(
    variants=lambkin.common.named_product(
        param_a=["x", "y"],
        param_b=[1, 10, 100],
    ),
    num_iterations=5,
)
@lambkin.option("--clock-rate", default=100.0)
def my_benchmark(ctx):
    with lambkin.process.background(ctx.shell.my_recorder, "-O", "output.mcap"):
        with lambkin.process.background(
            ctx.shell.my_algorithm,
            f"param_a:={ctx.variant.param_a}",
            f"param_b:={ctx.variant.param_b}",
            f"input:={ctx.inputs.dataset}",
        ):
            ctx.shell.my_player(ctx.inputs.dataset, "-r", ctx.options.clock_rate)

@my_benchmark.input
def dataset(ctx):
    return ctx.source.path.parent / "datasets" / "my_dataset.mcap"

> 

Reviewer: Consider adding the main as well for completion.

Author: Solved in d4a8d4b

```
> **Note:** Real execution is not yet supported. `--dry-run` is the only supported mode at this time. The shell proxy prints commands rather than running them.
> Background process orchestration will be implemented in subsequent phases.

For a complete, working example using the Beluga algorithm, see [`Beluga Example`](examples/beluga/beluga_benchmark.py).

## Requirements

- Python 3.10+
- [`uv`](https://github.com/astral-sh/uv)

## Installation

```bash
git clone git@github.com:Ekumen-OS/lambkin.git
uv sync

> ```

Reviewer: Consider moving them above the Usage section, as it is unusual to have them at the bottom.

Author: Solved in ae67ef5

Each stage is a Python callable that receives a context object carrying configuration, paths, and state. You implement the logic; LAMBKIN handles the rest.

### Core Concepts

LAMBKIN exposes a small set of composable primitives. Together they cover the full lifecycle of a benchmark — from declaring inputs and sweeping parameters to launching processes and collecting results.

**`named_product`**
Takes named parameter lists and returns every possible combination as a list of dictionaries, one per benchmark run configuration. Pass the result to `@benchmark` via `variants=` to sweep all combinations automatically.

**`@benchmark`**

> Drives the benchmark execution loop. It parses CLI options registered by `@option` once before the loop, then creates a `Context` for each `(variant, iteration)` pair and calls the decorated function with it. Input hooks registered via `@input` are resolved before each call and injected into `ctx.inputs`. Raises `ValueError` if `variants` is empty.

Reviewer: "then creates a `Context` for each `(variant, iteration)` pair" I think it is more accurate to describe it as It creates a Context for every loop.  Consider mentioning the base context used for the input registration.

Reviewer: "Raises ValueError if variants is empty" reads like internal implementation notes leaked into user docs. Worth dropping.

Author: Solved in 361c64a

**`Context`**
Carries all namespaced information for one benchmark variant. From the given parameters it builds `ctx.variant`, `ctx.inputs`, `ctx.options`, and `ctx.output`, and automatically creates the required output folders on disk before the benchmark function runs.

**`Source`**
Describes the benchmark script being executed. Exposed on the context as `ctx.source`, it gives benchmark stages access to the script's location and metadata without hardcoding paths.

**`@input`**

> Registers a hook on the benchmark function. Hooks are resolved before each call and their return values injected into `ctx.inputs` under the hook's function name, keeping data resolution decoupled from benchmark logic.

Reviewer: Same for the context accessor. `@input` should be more like `Input` and then in the usage we can see that can be used via decorator syntatic sugar. Same for the benchmark.

Author: Solved in 361c64a

**`@option`**

> Modelled after `click.option`. Registers a CLI option on the benchmark function by storing a `click.Option` object on `fn.__lambkin_options__`, which `@benchmark` collects and parses before the execution loop. The parsed values are injected into `ctx.options` — flag names are normalized by click, so `--clock-rate` becomes `ctx.options.clock_rate`.

Reviewer: Reference Click pkg. Same comment regarding the context. I think there is no need to be that specific regarding  names `fn.__lambkin_options__`.

Author: Solved in 361c64a

**`ctx.shell`**

> Exposes the host environment's executables as Python attributes. Accessing `ctx.shell.my_tool` returns a callable that, when invoked, runs `my_tool` with the given arguments. This lets benchmark scripts call external processes as if they were native Python functions, without hardcoding paths or constructing subprocess calls manually.

Reviewer: I think you are mixing concepts here.  This should not be ctx.shell, this should be Shell, which is an object or core element of the SDK per se. The fact that you access it via the Context is an architecture note focused on the usage.

Author: solved in 361c64a
